### PR TITLE
PythonAPI: Fix segfaults due to incorrect GIL locking

### PR DESCRIPTION
**Pull request by wentasah**
When using CARLA with Python 3.10, I'm getting a segfault in GetAvailableMaps. The problem disappears when PyList manipulation does not happen with GIL unlocked, as done in this commit.

The initial part of crash backtrace (from GDB) is below:

    Program terminated with signal SIGSEGV, Segmentation fault.

    warning: Section `.reg-xstate/49253' in core file too small.
    #0  _PyInterpreterState_GET () at ./Include/internal/pycore_pystate.h:117
    117	   return tstate->interp;
    [Current thread is 1 (Thread 0x7fe6fe48f740 (LWP 49253))]
    (gdb) bt
    #0  _PyInterpreterState_GET () at ./Include/internal/pycore_pystate.h:117
    #1  get_list_state () at Objects/listobject.c:26
    #2  PyList_New (size=0) at Objects/listobject.c:159
    #3  0x00007fe6fdc0dab0 in boost::python::detail::list_base::list_base() () from /nix/store/c95f3nrkz3sflvycihyw1c8q4nk47p4m-boost-1.79.0/lib/libboost_python310.so.1.79.0
    #4  0x00007fe6ef9ecfc4 in boost::python::list::list (this=0x7ffd8a8aae28) at include/boost/python/list.hpp:61
    #5  GetAvailableMaps (self=...) at source/libcarla/Client.cpp:26
    #6  0x00007fe6efb6a8fe in boost::python::detail::invoke<boost::python::to_python_value<boost::python::list const&>, boost::python::list (*)(carla::client::Client const&), boost::python::arg_from_python<carla::client::Client const&> > (rc=..., f=<optimized out>, ac0=...)
        at include/boost/python/detail/invoke.hpp:73
    #7  boost::python::detail::caller_arity<1u>::impl<boost::python::list (*)(carla::client::Client const&), boost::python::default_call_policies, boost::mpl::vector2<boost::python::list, carla::client::Client const&> >::operator() (args_=<optimized out>, this=<optimized out>)
        at include/boost/python/detail/caller.hpp:233
    #8  boost::python::objects::caller_py_function_impl<boost::python::detail::caller<boost::python::list (*)(carla::client::Client const&), boost::python::default_call_policies, boost::mpl::vector2<boost::python::list, carla::client::Client const&> > >::operator() (
        this=<optimized out>, args=<optimized out>, kw=<optimized out>) at include/boost/python/object/py_function.hpp:38
    #9  0x00007fe6fdc1b4dd in boost::python::objects::function::call(_object*, _object*) const () from /nix/store/c95f3nrkz3sflvycihyw1c8q4nk47p4m-boost-1.79.0/lib/libboost_python310.so.1.79.0
    #10 0x00007fe6fdc1b6a8 in boost::detail::function::void_function_ref_invoker0<boost::python::objects::(anonymous namespace)::bind_return, void>::invoke(boost::detail::function::function_buffer&) ()
       from /nix/store/c95f3nrkz3sflvycihyw1c8q4nk47p4m-boost-1.79.0/lib/libboost_python310.so.1.79.0
    ...

#### Description

Fixes #  

#### Where has this been tested?

  * **Platform(s):** Linux
  * **Python version(s):** 3.10
  * **Unreal Engine version(s):** default in 0.9.14

#### Possible Drawbacks

- - -
This change is [Reviewable](https://reviewable.io/reviews/carla-simulator/carla/6718)

### CODE REVIEW

**Comment by wentasah** [comment]
Added a changelog entry and rebased to `dev`.

File: PythonAPI/carla/source/libcarla/Client.cpp [hunk 1]
```diff
@@ -22,9 +22,13 @@ static void SetTimeout(carla::client::Client &client, double seconds) {
 }
 
 static auto GetAvailableMaps(const carla::client::Client &self) {
-  carla::PythonUtil::ReleaseGIL unlock;
   boost::python::list result;
-  for (const auto &str : self.GetAvailableMaps()) {
+  std::vector<std::string> maps;
+  {
+    carla::PythonUtil::ReleaseGIL unlock;
+    maps = self.GetAvailableMaps();
```
Comment (Harin22) [inline PythonAPI/carla/source/libcarla/Client.cpp:29]: Carla.fix

Comment (wentasah) [reply PythonAPI/carla/source/libcarla/Client.cpp:29]: What do you mean by this?

**Review by javiergrCS — DISMISSED** [review]
The code changes proposed by the collaborator wentasah have been successfully tested, verifying that they fix the segfault affecting the GetAvailableMaps function when using Python 3.10. The changes are ready to be merged and integrated into the 'dev' branch.

**Comment by wentasah** [comment]
Thanks for the review @javiergrCS.

In the mean time, I've found a similar segfault in `get_random_location_from_navigation()`. I added a commit fixing this segfault too.

**Review by javiergrCS — APPROVED** [review]
The changes proposed by the collaborator wentasah for the function get_random_location_from_navigation() have been successfully tested, and it has been verified that these changes fix the issue with that function. The changes are ready to be merged into the 'dev' branch.

Thank you for your collaboration, wentasah.

**Review by bernatx — APPROVED** [review]
> Reviewed 1 of 1 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/carla-simulator/carla/6718)* status: all files reviewed, 1 unresolved discussion (waiting on @Harin22)

**Review by bernatx — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/carla-simulator/carla/6718)* status: all files reviewed, 1 unresolved discussion (waiting on @Harin22)

**Pull request merged**